test/unit/get-shard-mapping.spec.js

```diff
@@ -0,0 +1,188 @@
+      }
+    });
+
+    it('should throw an error when utils.getUrl fails', async () => {
+      sinon.stub(utils, 'getDbs').resolves(['medic', 'medic-sentinel', 'medic-logs']);
+      sinon.stub(utils, 'getUrl').rejects(new Error('Failed to get URL'));
+
+      try {
+        await get_shard_mapping();
+        expect.fail('Should have thrown an error');
+      } catch (err) {
+        expect(err.message).to.equal('Failed to get shard mapping');
+        expect(consoleErrorStub.calledOnce).to.be.true;
+        expect(consoleErrorStub.firstCall.args[0]).to.equal('Error getting shard mapping:');
+      }
+    });
+
+    it('should throw an error when utils.request fails', async () => {
+      sinon.stub(utils, 'getDbs').resolves(['medic', 'medic-sentinel', 'medic-logs']);
+      sinon.stub(utils, 'getUrl').resolves('http://admin:pass@couchdb-1.local:5984/medic/_shards');
+      sinon.stub(utils, 'request').rejects(new Error('Failed to get shard info'));
+
+      try {
+        await get_shard_mapping();
+        expect.fail('Should have thrown an error');
+      } catch (err) {
+        expect(err.message).to.equal('Failed to get shard mapping');
+        expect(consoleErrorStub.calledOnce).to.be.true;
+        expect(consoleErrorStub.firstCall.args[0]).to.equal('Error getting shard mapping:');
+      }
+    });
+  });
+
+  describe('functionalities', () => {
+    it('should handle multiple databases with 8 shards and 3 nodes', async () => {
+      sinon.stub(utils, 'getDbs').resolves(['medic', 'medic-sentinel', 'medic-logs']);
+      sinon.stub(utils, 'getUrl')
+        .onCall(0).resolves('http://admin:pass@couchdb-1.local:5984/medic/_shards')
+        .onCall(1).resolves('http://admin:pass@couchdb-1.local:5984/medic-sentinel/_shards')
+        .onCall(2).resolves('http://admin:pass@couchdb-1.local:5984/medic-logs/_shards');
+
+      const shardInfo = {
+        shards: {
+          '00000000-1fffffff': ['node1'],
+          '20000000-3fffffff': ['node2'],
+          '40000000-5fffffff': ['node3'],
+          '60000000-7fffffff': ['node1'],
+          '80000000-9fffffff': ['node2'],
+          'a0000000-bfffffff': ['node3'],
+          'c0000000-dfffffff': ['node1'],
+          'e0000000-ffffffff': ['node2']
+        }
+      };
+
+      sinon.stub(utils, 'request').resolves(shardInfo);
+
+      const result = await get_shard_mapping();
+      const expectedResult = JSON.stringify({
+        '00000000-1fffffff': 'node1',
+        '20000000-3fffffff': 'node2',
+        '40000000-5fffffff': 'node3',
+        '60000000-7fffffff': 'node1',
+        '80000000-9fffffff': 'node2',
+        'a0000000-bfffffff': 'node3',
+        'c0000000-dfffffff': 'node1',
+        'e0000000-ffffffff': 'node2'
+      });
+
+      expect(result).to.equal(expectedResult);
+      expect(utils.getDbs.calledOnce).to.be.true;
+      expect(utils.getUrl.calledThrice).to.be.true;
+      expect(utils.request.calledThrice).to.be.true;
+    });
```

Comment:
Can you please add a unit test where a shard for one db is on one node, while for the other db the same shard is on another node? It'd be cool to see what happens in real life when this is the case. I've seen this occur very frequently.

Reply:
Thanks! This was enlightening. While digging into this, I investigated several clusters and it indeed turns out that the mapping for different databases are different. Some use 1, 2, 3 while others might use 2, 3, 1 and so forth. I've also confirmed that this is expected behavior for couch. One of the determinants for this is the timing of the database creation. Therefore this has necessitated changing the mapping to include database information - as there is no concept of a global shard map for all databases in the couch server. I have added the tests to confirm that the new approach works for this scenario. It's actually the only approach that works if you want to preserve shard mapping across migrations.